File: Java/LV3/야근 지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        long answer = 0;
        
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        for(int i=0;i<works.length;i++){
            pq.add(works[i]);
        }
        
        while(true){
            int top = pq.peek();
            if(top==0 || n==0){
                break;
            }
            
            else{
                n--;
                top-=1;
                pq.poll();
                pq.add(top);
            }
        }
        
        while(!pq.isEmpty()){
            answer+=(int)Math.pow(pq.peek(),2);
            pq.poll();
        }
        return answer;
    }
}

File: Lv.3/야근 지수.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
#include <functional>
#include <map>
#include <queue>
#include <math.h>
using namespace std;

long long solution(int n, vector<int> works) {
    long long answer = 0;
    priority_queue<int,vector<int>,less<int>> pq; // 내림차순 우선순위큐
    
    for(int i=0;i<works.size();i++){
        pq.push(works[i]);
    }
    
    while(1){
        int top = pq.top();
        if(top==0 || n==0){
            break;
        }
        else{
            n--;
            pq.pop();
            pq.push(top-1);
        }
    }
    
    while(!pq.empty()){
        answer+=(int)pow(pq.top(),2);
        pq.pop();
    }
    return answer;
}
